python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,395 @@
+
+                return mname, self.name_
+
+            def get_owner_indx(self):
+                return self.owner_.indx_
+
+            def get_dependent_str(self):
+                name = ""
+                for dependent in self.dependent_:
+                    mname, dname = dependent.get_name()
+                    name = name + (mname + ":output(" + dname if self.itype_ == 2 else "")
+                    name = name + (")" if self.itype_ == 2 else mname + ":" + dname)
+                return name
+
+            def add_dependent(self, dependent):
+                """
+                # check if the dependency setting correct.
+                # correct connection are following
+                # 1. global input to module input
+                # 2. module output to next module input
+                # 3. module output to global output
+                """
+                owner_indx = self.get_owner_indx()
+                dep_owner_indx = dependent.get_owner_indx()
+                assert owner_indx != dep_owner_indx, f"can not set self as dependent."
+                assert not (
+                    owner_indx > dep_owner_indx
+                    and not (dependent.itype_ == 2 and dep_owner_indx == 0)
+                ), f"dependent only can be next module interface or global output."
+                assert not (
+                    owner_indx == 0 and dependent.itype_ != 1
+                ), f"global input only can set dependent with module input."
+
+                self.dependent_.append(dependent)
+
+        def __init__(self, indx=0):
+            self.indx_ = indx
+            self.name_ = "mod" + str(indx) if indx else ""
+            self.interfaces_ = {1: {}, 2: {}}
+            self.target_host_ = None
+            self.mod_name_ = "default"
+            self.build_func_ = None
+            self.params_ = None
+            self.target_ = None
+            self.dev_ = None
+
+
+        def get_interface(self, itype, name):
+            if name not in self.interfaces_[itype]:
+                self.interfaces_[itype][name] = self.Interface(self, itype, name)
+
+            return self.interfaces_[itype][name]
+
+        def input(self, name):
+            return self.get_interface(1, name)
+
+        def output(self, index):
+            return self.get_interface(2, index)
+
+        def set_target_host(self, host):
+            self.target_host_ = host
+
+        def set_mod_name(self, name):
+            self.mod_name_ = name
+
+        def set_build_func(self, build_func):
+            self.build_func_ = build_func
+
+        def set_params(self, params):
+            self.params_ = params
+
+        def set_target(self, target):
+            self.target_ = target
+
+        def set_dev(self, dev):
+            self.dev_ = dev
+
+    def __init__(self, mods):
+        self.pipe_module_name_ = "pipeline_module"
+        self.mod_wrapper = {m: self.ModuleWrapper(i + 1) for m, i in zip(mods, range(len(mods)))}
```

Comment:
I feel this is a bit weird. It means when users specify `PipelineModuleConfig([mod0, mod1])`, they already specify the order. However, they still need to specify the connections later on. It would be better to have something like the following:

```
config = PipelineModuleConfig()
config[mod1].input("data0", config.pipeline_input("data0"))
config[mod2].input("data0", config.pipeline_input("data1"))
config[mod2].input("data1", config[mod2].output(0))
config[mod2].output(0, config.pipeline_output(0))
```

When `__getitem__` in pipeline module config sees a new module, it initializes its ModuleWrapper.

The side effect of this approach is that we cannot check if a connection is legal when adding a new connection. However, we can still check that later when exporting the config by leveraging topological sort.

Reply:
add the toplogy sort logic, the new API like following
```
pipe_config = pipeline_executor.PipelineConfig()
pipe_config["input"]["data_1"].connect(pipe_config[mod2]["input"]["data_1"])
pipe_config["input"]["data_1"].connect(pipe_config[mod2]["input"]["data_1"])
pipe_config[mod1]["output"][0].connect(pipe_config[mod2]["input"]["data_0"])
```